{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Init\n",
    "import os\n",
    "import sys\n",
    "\n",
    "module_path = os.path.abspath(os.path.join(\"../src/simulicronalpha/\"))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "# Imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import warnings\n",
    "import multiprocessing\n",
    "np.set_printoptions(suppress=True)\n",
    "from numpy import concatenate as c\n",
    "from numpy import cumsum\n",
    "import random\n",
    "%load_ext line_profiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from generateSim import (generatePopulation, generateGenome)\n",
    "from stats import stats\n",
    "from regulation import regulation\n",
    "# from checkCopyNumber import checkCopyNumber\n",
    "from transposition import transposition\n",
    "#from recombination import recombination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from popSim import runBatch, runSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gen,piset,piIndice, rates = generateGenome(\n",
    "    numberOfInsertionSites=10000,\n",
    "    numberOfChromosomes=6,\n",
    "    baseRecombinationRate=0.1,\n",
    "    baseSelection=0,\n",
    "    baseTau=1,\n",
    ")\n",
    "pop, tr , TEset= generatePopulation(\n",
    "    gen,\n",
    "    piIndice,\n",
    "    NumberOfIndividual=1000,\n",
    "    NumberOfTransposonTypes=1,\n",
    "    NumberOfInsertionsPerType=[1],\n",
    "    FrequencyOfInsertions=[1.0],\n",
    "    ExcisionRates = [1.0]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "*** KeyboardInterrupt exception caught in code being profiled."
     ]
    },
    {
     "data": {
      "text/plain": [
       "Timer unit: 1e-06 s\n",
       "\n",
       "Total time: 159.274 s\n",
       "File: /home/siddharth/Documents/Projects/Simulicron/src/simulicronalpha/popSim.py\n",
       "Function: runSim at line 25\n",
       "\n",
       "Line #      Hits         Time  Per Hit   % Time  Line Contents\n",
       "==============================================================\n",
       "    25                                           def runSim(\n",
       "    26                                               genomeMatrix,\n",
       "    27                                               populationMatrix,\n",
       "    28                                               transposonMatrix,\n",
       "    29                                               TEset,\n",
       "    30                                               NumberOfTransposonInsertions,\n",
       "    31                                               generations,\n",
       "    32                                               genMap,\n",
       "    33                                               piSet,\n",
       "    34                                           ):\n",
       "    35                                               # ------------------#\n",
       "    36                                               # lambda/macros\n",
       "    37         1          2.0      2.0      0.0      flatten = lambda *n: (\n",
       "    38                                                   e\n",
       "    39                                                   for a in n\n",
       "    40                                                   for e in (\n",
       "    41                                                       flatten(*a) if isinstance(a, (tuple, list)) else (a,)\n",
       "    42                                                   )\n",
       "    43                                               )\n",
       "    44                                               # ------------------#\n",
       "    45                                               # ------------------#\n",
       "    46                                               # for storing transposons which are fixed\n",
       "    47         1          1.0      1.0      0.0      fixedTE = []\n",
       "    48                                               # for storing transposons which are not fixed\n",
       "    49         1          1.0      1.0      0.0      unfixedTE = []\n",
       "    50                                               # for storing transposons which are lost\n",
       "    51         1          0.0      0.0      0.0      lostTE = []\n",
       "    52                                               # For storing the average copy number per generation\n",
       "    53         1          1.0      1.0      0.0      averageCopyNumber = []\n",
       "    54                                               # For storing the copy number variance per generation\n",
       "    55         1          1.0      1.0      0.0      varianceCopyNumber = []\n",
       "    56                                               # Per family statistics\n",
       "    57         1          1.0      1.0      0.0      TEfamilyCountArr = []\n",
       "    58         1          1.0      1.0      0.0      TEfamilyVarArr = []\n",
       "    59                                               # Create a insertionSiteSepecific array\n",
       "    60         1          1.0      1.0      0.0      insertionSiteFrequencyArray = np.zeros(\n",
       "    61         1          8.0      8.0      0.0          (len(genomeMatrix), NumberOfTransposonInsertions,)\n",
       "    62                                               )\n",
       "    63         1          1.0      1.0      0.0      transposonMatrixCopy = transposonMatrix\n",
       "    64         1          1.0      1.0      0.0      populationMatrixCopy = populationMatrix\n",
       "    65                                           \n",
       "    66                                               # Create a fixed size array for storing transposon information\n",
       "    67                                               # Also add a \"counter\" - with a large fixed array, it is not\n",
       "    68                                               # possible to use len(transposons) to find the last filled index\n",
       "    69         1          1.0      1.0      0.0      transposonMatrixCopy = np.append(\n",
       "    70         1          1.0      1.0      0.0          transposonMatrix,\n",
       "    71         1      96021.0  96021.0      0.1          np.zeros((10000000, 6), dtype=object),\n",
       "    72         1     320641.0 320641.0      0.2          axis=0,\n",
       "    73                                               )\n",
       "    74         1          4.0      4.0      0.0      numberOfTranspositionEvents = len(transposonMatrix)\n",
       "    75                                           \n",
       "    76                                               # Calculate piRNA coordinates\n",
       "    77         1          1.0      1.0      0.0      piCoord = []\n",
       "    78         7          7.0      1.0      0.0      for i in piSet.values():\n",
       "    79         6         18.0      3.0      0.0          piCoord = piCoord + (list(range(i[0], i[1])))\n",
       "    80                                           \n",
       "    81                                               # Calculate the CN and CNV for generation 0\n",
       "    82                                               (\n",
       "    83                                                   copyNumber,\n",
       "    84                                                   varianceNumber,\n",
       "    85                                                   TEfamilyCount,\n",
       "    86                                                   TEfamilyVar,\n",
       "    87         1          1.0      1.0      0.0      ) = checkCopyNumber(\n",
       "    88         1       4433.0   4433.0      0.0          populationMatrixCopy, TEset, transposonMatrixCopy\n",
       "    89                                               )\n",
       "    90         1          2.0      2.0      0.0      averageCopyNumber.append(copyNumber)\n",
       "    91         1          1.0      1.0      0.0      varianceCopyNumber.append(varianceNumber)\n",
       "    92         1          2.0      2.0      0.0      TEfamilyCountArr.append(TEfamilyCount)\n",
       "    93         1          2.0      2.0      0.0      TEfamilyVarArr.append(TEfamilyVar)\n",
       "    94                                           \n",
       "    95                                               # Driver loop\n",
       "    96        97        111.0      1.1      0.0      for i in range(generations):\n",
       "    97        97       6088.0     62.8      0.0          print(i)\n",
       "    98        97        417.0      4.3      0.0          populationV1 = []\n",
       "    99        97        339.0      3.5      0.0          populationV2 = []\n",
       "   100        97        655.0      6.8      0.0          populationFit = []\n",
       "   101     96429     104672.0      1.1      0.1          for k in list(range(populationMatrixCopy.shape[0])):\n",
       "   102     96333    1939767.0     20.1      1.2              fitness = list(populationMatrixCopy[0:, 2])\n",
       "   103                                           \n",
       "   104     96333     117142.0      1.2      0.1              p1, p2 = random.choices(\n",
       "   105     96333     862826.0      9.0      0.5                  list(range(populationMatrixCopy.shape[0])),\n",
       "   106     96333     100409.0      1.0      0.1                  weights=fitness,\n",
       "   107     96333    8739561.0     90.7      5.5                  k=2,\n",
       "   108                                                       )\n",
       "   109                                           \n",
       "   110                                                       # Since recombination function only accepts arrays,\n",
       "   111                                                       # checking and forcing type conversion as needed\n",
       "   112                                                       # for the respective alleles\n",
       "   113                                                       if (\n",
       "   114     96333     159063.0      1.7      0.1                  populationMatrixCopy[p1, 0] == 0\n",
       "   115      1660       1841.0      1.1      0.0                  and populationMatrixCopy[p1, 1] == 0\n",
       "   116                                                       ):\n",
       "   117      1117       1060.0      0.9      0.0                  v1 = 0\n",
       "   118                                                       else:\n",
       "   119     95216     128616.0      1.4      0.1                  cP1V1 = populationMatrixCopy[p1, 0]\n",
       "   120     95216     117032.0      1.2      0.1                  cP1V2 = populationMatrixCopy[p1, 1]\n",
       "   121     95216     121422.0      1.3      0.1                  if isinstance(cP1V1, list):\n",
       "   122     94673     770626.0      8.1      0.5                      cP1V1 = np.asarray(cP1V1)\n",
       "   123                                                           else:\n",
       "   124       543       2380.0      4.4      0.0                      cP1V1 = np.asarray([cP1V1])\n",
       "   125     95216     106970.0      1.1      0.1                  if isinstance(cP1V2, list):\n",
       "   126     94688     569710.0      6.0      0.4                      cP1V2 = np.asarray(cP1V2)\n",
       "   127                                                           else:\n",
       "   128       528       1362.0      2.6      0.0                      cP1V2 = np.asarray([cP1V2])\n",
       "   129                                           \n",
       "   130     95216      92636.0      1.0      0.1                  v1 = recombination(\n",
       "   131     95216   56931401.0    597.9     35.7                      genMap, transposonMatrixCopy, v1=cP1V1, v2=cP1V2,\n",
       "   132                                                           )\n",
       "   133                                           \n",
       "   134                                                       if (\n",
       "   135     96332     177411.0      1.8      0.1                  populationMatrixCopy[p2, 0] == 0\n",
       "   136      1638       1798.0      1.1      0.0                  and populationMatrixCopy[p2, 1] == 0\n",
       "   137                                                       ):\n",
       "   138      1103       1050.0      1.0      0.0                  v2 = 0\n",
       "   139                                                       else:\n",
       "   140     95229     127733.0      1.3      0.1                  cP2V1 = populationMatrixCopy[p2, 0]\n",
       "   141     95229     118069.0      1.2      0.1                  cP2V2 = populationMatrixCopy[p2, 1]\n",
       "   142     95229     118001.0      1.2      0.1                  if isinstance(cP2V1, list):\n",
       "   143     94694     724202.0      7.6      0.5                      cP2V1 = np.asarray(cP2V1)\n",
       "   144                                                           else:\n",
       "   145       535       1922.0      3.6      0.0                      cP2V1 = np.asarray([cP2V1])\n",
       "   146     95229     107183.0      1.1      0.1                  if isinstance(cP2V2, list):\n",
       "   147     94683     571253.0      6.0      0.4                      cP2V2 = np.asarray(cP2V2)\n",
       "   148                                                           else:\n",
       "   149       546       1332.0      2.4      0.0                      cP2V2 = np.asarray([cP2V2])\n",
       "   150                                           \n",
       "   151     95229      94213.0      1.0      0.1                  v2 = recombination(\n",
       "   152     95229   56501068.0    593.3     35.5                      genMap, transposonMatrixCopy, v1=cP2V1, v2=cP2V2,\n",
       "   153                                                           )\n",
       "   154                                           \n",
       "   155     96332     150967.0      1.6      0.1              if v1 == 0 and v2 == 0:\n",
       "   156       737        720.0      1.0      0.0                  indFitness = 1\n",
       "   157                                           \n",
       "   158                                                       else:\n",
       "   159                                                           (\n",
       "   160                                                               v1,\n",
       "   161                                                               v2,\n",
       "   162                                                               transposonMatrixCopy,\n",
       "   163                                                               TEset,\n",
       "   164                                                               numberOfTranspositionEvents,\n",
       "   165     95595     108269.0      1.1      0.1                  ) = transposition(\n",
       "   166     95595      94814.0      1.0      0.1                      transposonMatrix=transposonMatrixCopy,\n",
       "   167     95595      92497.0      1.0      0.1                      genomeMatrix=genomeMatrix,\n",
       "   168     95595      89904.0      0.9      0.1                      NumberOfTransposonInsertions=NumberOfTransposonInsertions,\n",
       "   169     95595      92082.0      1.0      0.1                      TEset=TEset,\n",
       "   170     95595      90953.0      1.0      0.1                      piCoord=piCoord,\n",
       "   171     95595      91886.0      1.0      0.1                      numberOfTranspositionEvents=numberOfTranspositionEvents,\n",
       "   172     95595      89738.0      0.9      0.1                      v1=v1,\n",
       "   173     95595   20443566.0    213.9     12.8                      v2=v2,\n",
       "   174                                                           )\n",
       "   175     95595     134425.0      1.4      0.1                  indFitness = calculateFitness(\n",
       "   176     95595    4044354.0     42.3      2.5                      transposonMatrixCopy, v1, v2\n",
       "   177                                                           )\n",
       "   178                                           \n",
       "   179     96332     135546.0      1.4      0.1              populationV1.append(v1)\n",
       "   180     96332     100580.0      1.0      0.1              populationV2.append(v2)\n",
       "   181     96332     101330.0      1.1      0.1              populationFit.append(indFitness)\n",
       "   182                                                   # Return i+2 since i start at 0 = generation 1\n",
       "   183                                                   # and the condition check happens at generation\n",
       "   184                                                   # n-1, hence i + 1 + 1\n",
       "   185                                           \n",
       "   186                                                   # Check if there are no transposons left\n",
       "   187        96         97.0      1.0      0.0          if all(\n",
       "   188        96        104.0      1.1      0.0              np.array_equal(v, [0, 0])\n",
       "   189        96    1282192.0  13356.2      0.8              for v in np.c_[populationV1, populationV2]\n",
       "   190                                                   ):\n",
       "   191                                                       return {\n",
       "   192                                                           \"State\": \"LOSS\",\n",
       "   193                                                           \"Generatrion\": i + 2,\n",
       "   194                                                           \"NTE\": numberOfTranspositionEvents,\n",
       "   195                                                           \"AvgCopyNum\": averageCopyNumber,\n",
       "   196                                                           \"CopyNumVar\": varianceCopyNumber,\n",
       "   197                                                           \"TEfamilyCN\": TEfamilyCountArr,\n",
       "   198                                                           \"TEfamilyVR\": TEfamilyVarArr,\n",
       "   199                                                       }\n",
       "   200                                                   else:\n",
       "   201                                                       pass\n",
       "   202                                                   # Major bug in numpy - forced to use pandas\n",
       "   203                                                   # Refer to the question\n",
       "   204                                                   # https://stackoverflow.com/questions/60210897\n",
       "   205                                                   # populationMatrixCopy = pd.DataFrame(\n",
       "   206                                                   #    [populationV1, populationV2, populationFit]\n",
       "   207                                                   # ).T.to_numpy()\n",
       "   208        96        177.0      1.8      0.0          populationMatrixCopy= np.array([populationV1,\n",
       "   209        96         93.0      1.0      0.0              populationV2,\n",
       "   210        96     792478.0   8255.0      0.5              populationFit,], dtype=\"object\").T\n",
       "   211                                                   # (\n",
       "   212                                                   #     compyNumber,\n",
       "   213                                                   #     varianceNumber,\n",
       "   214                                                   #     insertionSiteFrequencyArray,\n",
       "   215                                                   # ) = statistics(\n",
       "   216                                                   #     populationMatrixCopy,\n",
       "   217                                                   #     transposonMatrixCopy,\n",
       "   218                                                   #     TEset,\n",
       "   219                                                   #     insertionSiteFrequencyArray,\n",
       "   220                                                   # )\n",
       "   221                                                   (\n",
       "   222                                                       copyNumber,\n",
       "   223                                                       varianceNumber,\n",
       "   224                                                       TEfamilyCount,\n",
       "   225                                                       TEfamilyVar,\n",
       "   226        96        208.0      2.2      0.0          ) = checkCopyNumber(\n",
       "   227        96    1494362.0  15566.3      0.9              populationMatrixCopy, TEset, transposonMatrixCopy\n",
       "   228                                                   )\n",
       "   229        96        171.0      1.8      0.0          averageCopyNumber.append(copyNumber)\n",
       "   230        96        110.0      1.1      0.0          varianceCopyNumber.append(varianceNumber)\n",
       "   231        96        101.0      1.1      0.0          TEfamilyCountArr.append(TEfamilyCount)\n",
       "   232        96        101.0      1.1      0.0          TEfamilyVarArr.append(TEfamilyVar)\n",
       "   233                                           \n",
       "   234                                               # Quit simulation if there in a transient state\n",
       "   235                                               # i.e. no loss\n",
       "   236                                               return {\n",
       "   237                                                   \"State\": \"FLUX\",\n",
       "   238                                                   \"Generatrion\": i + 2,\n",
       "   239                                                   \"NTE\": numberOfTranspositionEvents,\n",
       "   240                                                   \"AvgCopyNum\": averageCopyNumber,\n",
       "   241                                                   \"CopyNumVar\": varianceCopyNumber,\n",
       "   242                                                   \"TEfamilyCN\": TEfamilyCountArr,\n",
       "   243                                                   \"TEfamilyVR\": TEfamilyVarArr,\n",
       "   244                                               }"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%lprun -f runSim runSim(gen,pop,tr,TEset,1,10000,rates,piset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "runSim(gen,pop,tr,TEset,1,100,rates,piset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%lprun -f recombination recombination(rates=rates, transposonMatrix=tr, v1=[52], v2=[65])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%lprun -f runBatch runBatch(numberOfSimulations=1,baseSelection=0,baseInsertionProb=1,numberOfInsertionSites=10000,numberOfChromosomes=6,baseRecombinationRate=0.01,baseTau=1,numberOfPiRNA=6,piPercentage=3,enablePiRecombination=False,NumberOfIndividual=1000,NumberOfTransposonTypes=1,NumberOfInsertionsPerType=[1],FrequencyOfInsertions=[1.0],ExcisionRates=[1.0],RepairRates=[1],InsertionRates=[1],HardyWeinberg=False,NumberOfGenerations=10,numberOfThreads=1,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.set_printoptions(suppress=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%lprun -f transposition transposition(transposonMatrix=tr, genomeMatrix=gen, NumberOfTransposonInsertions=1, TEset=TEset, v1=[52,101,321,433,12,23,234,123,657,29], v2=[65,54,22,121,23])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def checkCopyNumber(populationMatrix, TEset, transposonMatrix):\n",
    "    # The local flatten lambda\n",
    "    flatten = lambda *n: (\n",
    "        e for a in n for e in (flatten(*a) if isinstance(a, (tuple, list)) else (a,))\n",
    "    )\n",
    "    # Convert the list to flat list and remove 0\n",
    "    # if present.\n",
    "    # Also create a dictionary counting \n",
    "    # each family type\n",
    "    TEfamilyCount = {k: [] for k in range(1, len(TEset.keys()) + 1)}\n",
    "    TEfamilyVar = {k: [] for k in range(1, len(TEset.keys()) + 1)}\n",
    "    TETotal = []\n",
    "    for i in range(len(populationMatrix)):\n",
    "        v1 = populationMatrix[i, 0] if type(populationMatrix[i, 0]) is list else []\n",
    "        v2 = populationMatrix[i, 1] if type(populationMatrix[i, 1]) is list else []\n",
    "        if (v1 + v2 == []):\n",
    "            TETotal.append(0)\n",
    "            for k in TEset.keys():\n",
    "                TEfamilyCount[k].append(0)\n",
    "        else:\n",
    "            # Check the occupancy for each family\n",
    "            TEfamilies = transposonMatrix[v1+v2, 0].tolist()\n",
    "            TETotal.append(len(TEfamilies))\n",
    "            for k in TEset.keys():\n",
    "                TEfamilyCount[k].append(TEfamilies.count(k))\n",
    "    # Calculate variance per family and sum the count\n",
    "    for k in TEset.keys():\n",
    "        TEfamilyVar[k].append(np.var(TEfamilyCount[k]))\n",
    "        TEfamilyCount[k] = sum(TEfamilyCount[k])\n",
    "    return sum(TETotal), np.var(TETotal), TEfamilyCount, TEfamilyVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TEfamilyCount = {k: [] for k in range(1, len(TEset.keys()) + 1)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%timeit\n",
    "x, y, a, b = checkCopyNumber(pop, TEset, tr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TEfamilyCount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 0, 1],\n",
       "       [list([682, 759]), 0, 1.0],\n",
       "       [0, 0, 1],\n",
       "       ...,\n",
       "       [list([401, 613]), 0, 1.0],\n",
       "       [0, list([62]), 1.0],\n",
       "       [0, 0, 1]], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
