{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Init\n",
    "import os\n",
    "import sys\n",
    "\n",
    "module_path = os.path.abspath(os.path.join(\"../src/simulicronalpha/\"))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "# Imports\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import warnings\n",
    "import multiprocessing\n",
    "np.set_printoptions(suppress=True)\n",
    "from numpy import concatenate as c\n",
    "from numpy import cumsum\n",
    "import random\n",
    "%load_ext line_profiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from generateSim import (generatePopulation, generateGenome)\n",
    "from stats import stats\n",
    "from regulation import regulation\n",
    "from checkCopyNumber import checkCopyNumber\n",
    "#from recombination import recombination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from popSim import runBatch, runSim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gen,piset,piIndice, rates = generateGenome(\n",
    "    numberOfInsertionSites=10000,\n",
    "    numberOfChromosomes=6,\n",
    "    baseRecombinationRate=0.1,\n",
    "    baseSelection=0,\n",
    ")\n",
    "pop, tr , TEset= generatePopulation(\n",
    "    gen,\n",
    "    piIndice,\n",
    "    NumberOfIndividual=1000,\n",
    "    NumberOfTransposonTypes=1,\n",
    "    NumberOfInsertionsPerType=[1],\n",
    "    FrequencyOfInsertions=[1.0],\n",
    "    ExcisionRates = [1.0]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Timer unit: 1e-06 s\n",
       "\n",
       "Total time: 279.298 s\n",
       "File: /home/siddharth/Documents/Projects/Simulicron/src/simulicronalpha/popSim.py\n",
       "Function: runSim at line 25\n",
       "\n",
       "Line #      Hits         Time  Per Hit   % Time  Line Contents\n",
       "==============================================================\n",
       "    25                                           def runSim(\n",
       "    26                                               genomeMatrix,\n",
       "    27                                               populationMatrix,\n",
       "    28                                               transposonMatrix,\n",
       "    29                                               TEset,\n",
       "    30                                               NumberOfTransposonInsertions,\n",
       "    31                                               generations,\n",
       "    32                                               genMap,\n",
       "    33                                           ):\n",
       "    34                                               # ------------------#\n",
       "    35                                               # lambda/macros\n",
       "    36         1          2.0      2.0      0.0      flatten = lambda *n: (\n",
       "    37                                                   e for a in n for e in (flatten(*a) if isinstance(a, (tuple, list)) else (a,))\n",
       "    38                                               )\n",
       "    39                                               # ------------------#\n",
       "    40                                               # ------------------#\n",
       "    41                                               # for storing transposons which are fixed\n",
       "    42         1          1.0      1.0      0.0      fixedTE = []\n",
       "    43                                               # for storing transposons which are not fixed\n",
       "    44         1          1.0      1.0      0.0      unfixedTE = []\n",
       "    45                                               # for storing transposons which are lost\n",
       "    46         1          1.0      1.0      0.0      lostTE = []\n",
       "    47                                               # For storing the average copy number per generation\n",
       "    48         1          1.0      1.0      0.0      averageCopyNumber = []\n",
       "    49                                               # For storing the copy number variance per generation\n",
       "    50         1          1.0      1.0      0.0      varianceCopyNumber = []\n",
       "    51                                               # Create a insertionSiteSepecific array\n",
       "    52         2         12.0      6.0      0.0      insertionSiteFrequencyArray = np.zeros(\n",
       "    53         1          1.0      1.0      0.0          (len(genomeMatrix), NumberOfTransposonInsertions,)\n",
       "    54                                               )\n",
       "    55         1          1.0      1.0      0.0      transposonMatrixCopy = transposonMatrix\n",
       "    56         1          1.0      1.0      0.0      populationMatrixCopy = populationMatrix\n",
       "    57                                           \n",
       "    58                                               # Calculate the CN and CNV for generation 0\n",
       "    59         1       9795.0   9795.0      0.0      copyNumber, varianceNumber = checkCopyNumber(populationMatrixCopy)\n",
       "    60         1          2.0      2.0      0.0      averageCopyNumber.append(copyNumber)\n",
       "    61         1          1.0      1.0      0.0      varianceCopyNumber.append(varianceNumber)\n",
       "    62                                           \n",
       "    63                                               # Driver loop\n",
       "    64        31         29.0      0.9      0.0      for i in range(generations):\n",
       "    65                                                   # print(i)\n",
       "    66        30        131.0      4.4      0.0          populationV1 = []\n",
       "    67        30         99.0      3.3      0.0          populationV2 = []\n",
       "    68        30        142.0      4.7      0.0          populationFit = []\n",
       "    69     30030      28000.0      0.9      0.0          for k in list(range(populationMatrixCopy.shape[0])):\n",
       "    70     30000     589543.0     19.7      0.2              fitness = list(populationMatrixCopy[0:, 2])\n",
       "    71                                           \n",
       "    72     60000    3023668.0     50.4      1.1              p1, p2 = random.choices(\n",
       "    73     30000     289842.0      9.7      0.1                  list(range(populationMatrixCopy.shape[0])), weights=fitness, k=2,\n",
       "    74                                                       )\n",
       "    75                                           \n",
       "    76                                                       # Since recombination function only accepts arrays,\n",
       "    77                                                       # checking and forcing type conversion as needed\n",
       "    78                                                       # for the respective alleles\n",
       "    79     30000      48412.0      1.6      0.0              if populationMatrixCopy[p1, 0] == 0 and populationMatrixCopy[p1, 1] == 0:\n",
       "    80       970        821.0      0.8      0.0                  v1 = 0\n",
       "    81                                                       else:\n",
       "    82     29030      35826.0      1.2      0.0                  cP1V1 = populationMatrixCopy[p1, 0]\n",
       "    83     29030      32956.0      1.1      0.0                  cP1V2 = populationMatrixCopy[p1, 1]\n",
       "    84     29030      33009.0      1.1      0.0                  if isinstance(cP1V1, list):\n",
       "    85     28513     177138.0      6.2      0.1                      cP1V1 = np.asarray(cP1V1)\n",
       "    86                                                           else:\n",
       "    87       517       2094.0      4.1      0.0                      cP1V1 = np.asarray([cP1V1])\n",
       "    88     29030      27873.0      1.0      0.0                  if isinstance(cP1V2, list):\n",
       "    89     28514     119386.0      4.2      0.0                      cP1V2 = np.asarray(cP1V2)\n",
       "    90                                                           else:\n",
       "    91       516       1242.0      2.4      0.0                      cP1V2 = np.asarray([cP1V2])\n",
       "    92                                           \n",
       "    93     29030   11315740.0    389.8      4.1                  v1 = recombination(genMap, transposonMatrixCopy, v1=cP1V1, v2=cP1V2,)\n",
       "    94                                           \n",
       "    95     30000      53081.0      1.8      0.0              if populationMatrixCopy[p2, 0] == 0 and populationMatrixCopy[p2, 1] == 0:\n",
       "    96       939        821.0      0.9      0.0                  v2 = 0\n",
       "    97                                                       else:\n",
       "    98     29061      36269.0      1.2      0.0                  cP2V1 = populationMatrixCopy[p2, 0]\n",
       "    99     29061      33817.0      1.2      0.0                  cP2V2 = populationMatrixCopy[p2, 1]\n",
       "   100     29061      33786.0      1.2      0.0                  if isinstance(cP2V1, list):\n",
       "   101     28571     160969.0      5.6      0.1                      cP2V1 = np.asarray(cP2V1)\n",
       "   102                                                           else:\n",
       "   103       490       1655.0      3.4      0.0                      cP2V1 = np.asarray([cP2V1])\n",
       "   104     29061      27645.0      1.0      0.0                  if isinstance(cP2V2, list):\n",
       "   105     28508     118047.0      4.1      0.0                      cP2V2 = np.asarray(cP2V2)\n",
       "   106                                                           else:\n",
       "   107       553       1285.0      2.3      0.0                      cP2V2 = np.asarray([cP2V2])\n",
       "   108                                           \n",
       "   109     29061   11033523.0    379.7      4.0                  v2 = recombination(genMap, transposonMatrixCopy, v1=cP2V1, v2=cP2V2,)\n",
       "   110                                           \n",
       "   111     30000      41695.0      1.4      0.0              if v1 == 0 and v2 == 0:\n",
       "   112       614        533.0      0.9      0.0                  indFitness = 1\n",
       "   113                                           \n",
       "   114                                                       else:\n",
       "   115     58772  243158070.0   4137.3     87.1                  v1, v2, transposonMatrixCopy, TEset = transposition(\n",
       "   116     29386      24145.0      0.8      0.0                      transposonMatrix=transposonMatrixCopy,\n",
       "   117     29386      23138.0      0.8      0.0                      genomeMatrix=genomeMatrix,\n",
       "   118     29386      22844.0      0.8      0.0                      NumberOfTransposonInsertions=NumberOfTransposonInsertions,\n",
       "   119     29386      23078.0      0.8      0.0                      TEset=TEset,\n",
       "   120     29386      22592.0      0.8      0.0                      v1=v1,\n",
       "   121     29386      22611.0      0.8      0.0                      v2=v2,\n",
       "   122                                                           )\n",
       "   123     29386    1008664.0     34.3      0.4                  indFitness = calculateFitness(transposonMatrixCopy, v1, v2)\n",
       "   124                                           \n",
       "   125     30000      39181.0      1.3      0.0              populationV1.append(v1)\n",
       "   126     30000      27400.0      0.9      0.0              populationV2.append(v2)\n",
       "   127     30000      26502.0      0.9      0.0              populationFit.append(indFitness)\n",
       "   128                                                   # Return i+2 since i start at 0 = generation 1\n",
       "   129                                                   # and the condition check happens at generation\n",
       "   130                                                   # n-1, hence i + 1 + 1\n",
       "   131                                           \n",
       "   132                                                   # Check if there are no transposons left\n",
       "   133        30     232255.0   7741.8      0.1          if all(np.array_equal(v, [0, 0]) for v in np.c_[populationV1, populationV2]):\n",
       "   134                                                       return {\n",
       "   135                                                           \"State\": \"LOSS\",\n",
       "   136                                                           \"Generatrion\": i + 2,\n",
       "   137                                                           \"NTE\": transposonMatrixCopy.size / 4 - 1,\n",
       "   138                                                           \"AvgCopyNum\": averageCopyNumber,\n",
       "   139                                                           \"CopyNumVar\": varianceCopyNumber,\n",
       "   140                                                       }\n",
       "   141                                                   else:\n",
       "   142                                                       pass\n",
       "   143                                                   # Major bug in numpy - forced to use pandas\n",
       "   144                                                   # Refer to the question\n",
       "   145                                                   # https://stackoverflow.com/questions/60210897\n",
       "   146        60    3637544.0  60625.7      1.3          populationMatrixCopy = pd.DataFrame(\n",
       "   147        30         29.0      1.0      0.0              [populationV1, populationV2, populationFit]\n",
       "   148                                                   ).T.to_numpy()\n",
       "   149                                                   # (\n",
       "   150                                                   #     compyNumber,\n",
       "   151                                                   #     varianceNumber,\n",
       "   152                                                   #     insertionSiteFrequencyArray,\n",
       "   153                                                   # ) = statistics(\n",
       "   154                                                   #     populationMatrixCopy,\n",
       "   155                                                   #     transposonMatrixCopy,\n",
       "   156                                                   #     TEset,\n",
       "   157                                                   #     insertionSiteFrequencyArray,\n",
       "   158                                                   # )\n",
       "   159        30    3750897.0 125029.9      1.3          copyNumber, varianceNumber = checkCopyNumber(populationMatrixCopy)\n",
       "   160        30         70.0      2.3      0.0          averageCopyNumber.append(copyNumber)\n",
       "   161        30         32.0      1.1      0.0          varianceCopyNumber.append(varianceNumber)\n",
       "   162                                               # Quit simulation if there in a transient state\n",
       "   163                                               # i.e. no loss\n",
       "   164         1          1.0      1.0      0.0      return {\n",
       "   165         1          1.0      1.0      0.0          \"State\": \"FLUX\",\n",
       "   166         1          1.0      1.0      0.0          \"Generatrion\": i + 2,\n",
       "   167         1          2.0      2.0      0.0          \"NTE\": transposonMatrixCopy.size / 4 - 1,\n",
       "   168         1          1.0      1.0      0.0          \"AvgCopyNum\": averageCopyNumber,\n",
       "   169         1          1.0      1.0      0.0          \"CopyNumVar\": varianceCopyNumber,\n",
       "   170                                               }"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%lprun -f runSim runSim(gen,pop,tr,TEset,1,30,rates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%lprun -f recombination recombination(rates=rates, transposonMatrix=tr, v1=[52], v2=[65])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%lprun -f runBatch runBatch(numberOfSimulations=1,baseSelection=0,baseInsertionProb=1,numberOfInsertionSites=10000,numberOfChromosomes=6,baseRecombinationRate=0.01,baseTau=1,numberOfPiRNA=6,piPercentage=3,enablePiRecombination=False,NumberOfIndividual=1000,NumberOfTransposonTypes=1,NumberOfInsertionsPerType=[1],FrequencyOfInsertions=[1.0],ExcisionRates=[1.0],RepairRates=[1],InsertionRates=[1],HardyWeinberg=False,NumberOfGenerations=10,numberOfThreads=1,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "recombination(rates=rates, transposonMatrix=tr, v1=[52,65], v2=[52,65])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.set_printoptions(suppress=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recombination(rates, transposonMatrix, v1, v2):\n",
    "    # Empty vectors to store result\n",
    "    r1 = []\n",
    "    r2 = []\n",
    "    # Creating lambda (macro)\n",
    "    # \"Match\" does not exist in python\n",
    "    # match = lambda a, b: [ b.index(x) if x in b else 0 for x in a ]\n",
    "    # Get the postion of transposons \n",
    "    positionV1 = transposonMatrix[v1,1].astype(int).tolist()\n",
    "    positionV2 = transposonMatrix[v2,1].astype(int).tolist()\n",
    "    # This step sorts the locations and adds another location, \n",
    "    # 0 if not already present\n",
    "    unqiquePos = list(set(positionV1+positionV2) | set([0]))\n",
    "    unqiquePos.sort()\n",
    "    # Calculate the effective rate from genome map\n",
    "    effectiveRates = 0.5*(1-np.exp(-2*np.diff(rates[unqiquePos])))\n",
    "    # print (effectiveRates)\n",
    "    # Performing \"Recombination\"\n",
    "    rec = (np.random.uniform(size=len(effectiveRates)) < effectiveRates)\n",
    "    # Select the direction to start from\n",
    "    start = [(np.random.uniform() < 0.5)]\n",
    "    # Concat. start and recombination\n",
    "    # Also remove the added 0 and start from\n",
    "    # whichhaplo and uniqpos\n",
    "    whichhaplo = 1 + cumsum(c((start, rec))) % 2\n",
    "    whichhaplo = np.delete(whichhaplo, 0)\n",
    "    del unqiquePos[0]\n",
    "    unqiquePos = np.asarray(unqiquePos)\n",
    "    # Generating the haplotype\n",
    "    # Also checking if there is no transposon left\n",
    "    # In the case above, return a (int) 0\n",
    "    # Else return the array containing transposons\n",
    "    if (positionV1 == [0]):\n",
    "        r1 = []\n",
    "    else:\n",
    "        if not any(whichhaplo == 1):\n",
    "            pass\n",
    "        else:\n",
    "            pos = unqiquePos[whichhaplo == 1]\n",
    "            for i in v1:\n",
    "                if (int(transposonMatrix[i,1]) in pos):\n",
    "                    r1.append(i)\n",
    "    if (positionV2 == [0]):\n",
    "        r2 = []\n",
    "    else:\n",
    "        if not any(whichhaplo == 2):\n",
    "            pass\n",
    "        else:\n",
    "            pos = unqiquePos[whichhaplo == 2]\n",
    "            for i in v2:\n",
    "                if (int(transposonMatrix[i,1]) in pos):\n",
    "                    r2.append(i)\n",
    "    # Merge to create gamate\n",
    "    r = r1 + r2\n",
    "    # Return 0 if no transposon remains\n",
    "    if not r:\n",
    "        return (0)\n",
    "    else:\n",
    "        return r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
